test/async-hooks/test-async-functions-context.js

```diff
@@ -6,12 +6,16 @@ const { AsyncContext } = require('async_hooks');
 async function foo() {}
 
 async function testAwait() {
-
-  const asyncContext = new AsyncContext();
-  asyncContext.enter().set("key", "value");
   await foo();
   assert.notStrictEqual(asyncContext.getStore(), undefined);  // asserts
   assert.strictEqual(asyncContext.getStore().get("key"), "value");
 }
 
-assert.doesNotReject(testAwait());
+const asyncContext = new AsyncContext();
+asyncContext.enter((store) => {
+  store.set("key", "value");
+  assert.doesNotReject(testAwait());
```

Comment:
Actually this tests something different now. I could imagine that the sync variant would work also if the value is set before `testAwait()` is called.
Additionally it shows the limitations of the async variant - users potentiality have to modify call tree of their apps.

Reply:
I don't believe the `assert.doesNotReject()` wrapper around `testAwait()` is actually necessary here. I'm fairly certain we now have it so that the test will fail if there is an unhandled rejection. /cc @BridgeAR

Reply:
Yes, the test would indeed fail in case an unhandled rejection occurs. Using `doesNotReject` would also keep it as rejection, so even if that would not fail, it would not help either. Thus I also opened #27402 to prevent any of these from being added to our code base.

Reply:
@Flarna actually for APM case, I would call the `enter` in a monkeypatch on the event emission of http.Server. This is how I work with the CLS currently. I am not sure what alternative test you would recommend with the async version here.

@jasnell @BridgeAR thanks for the heads up, will update.

Reply:
@vdeturckheim I agree that async version most likely works for e.g. incoming HTTP requests which are anyway async already before. But There are other usecases like a sync call to `someDb.query()` where breaking the sync flow changes application behavior.
I expect similar issues with `exit()` to stop tracking an async tree - or just one branch of a tree.

Reply:
@Flarna I did not find any occurence of such instrumentation in main APMs and RASP tools. Do you have an example of such database?

Reply:
@vdeturckheim Not sure if I understand your question correct. For example mongodb `Collection.findAndModify()` will return a promise therefore it can be awaited. every APM tool which wants to track this call has to keep context from call location of `findAndModify()` till the promise get's resolved (or the callback is called).
Currently this is usually done by wrapping `findAndModify` and the wrapper first stores it's context, calls the original `findAndModify` and then it clears it's context.

Reply:
Gotcha! I was missunderstanding what you called a "sync call" here.

Reply:
As indicated in some of my other comments I was also working on something similar then this here. Meanwhile I created a #27172 to allow also other to take a look and comment independent of your PR. 
Not sure which variant is better nor if we can combine both. But I would love to get some comment from your side on it.

Reply:
Sorry for the late answer, I'll try to review your PR. tbh, the main goal of this current PR is to provide a memory-leak free API to do this and inspired from the CLS somehow (which set the tone for years in the ecosystem).